pkgs/desktops/gnome-3/core/gnome-session/default.nix

```diff
@@ -48,6 +50,14 @@ stdenv.mkDerivation rec {
       --suffix XDG_DATA_DIRS : "$out/share:$GSETTINGS_SCHEMAS_PATH" \
       --suffix XDG_DATA_DIRS : "${gnome3.gnome-shell}/share"\
       --suffix XDG_CONFIG_DIRS : "${gnome3.gnome-settings-daemon}/etc/xdg"
+  ''
+  # We hide the GNOME sessions by default since gnome-session is a dependency of
+  # GDM itself. If we do not hide them by default, it will show broken GNOME sessions
+  # when GDM is enable without proper GNOME installation.
+  + stdenv.lib.optionalString hiddenSessions ''
```

Comment:
It might be better to move the desktop files to their own special output, eg. `gnome3.gnome-session.sessions`, this can then be used in `sessionPackages`. That way we won't need to duplicate the whole derivation. 

We could also simply remove `${gnome-session}/share` from the gnome-session wrapper[0](https://github.com/NixOS/nixpkgs/blob/master/pkgs/desktops/gnome-3/core/gnome-session/default.nix#L36). It's there so gnome-session can find its own `share/gnome-session/sessions/gnome.session`, which is now only necessary for the x11 fallback. Though I'm guessing it might be best to keep it for now as the fallback can be useful.

Reply:
Thank you @hedning for taking time to review !

> It might be better to move the desktop files to their own special output, eg. `gnome3.gnome-session.sessions`, this can then be used in `sessionPackages`. That way we won't need to duplicate the whole derivation.

I like this idea, I'll try to find an example of another package on `nixpkgs` to learn how to make the split. (If you have an idea on which one I could take reference from, that would be great).

> We could also simply remove `${gnome-session}/share` from the gnome-session wrapper[0](https://github.com/NixOS/nixpkgs/blob/master/pkgs/desktops/gnome-3/core/gnome-session/default.nix#L36). It's there so gnome-session can find its own `share/gnome-session/sessions/gnome.session`, which is now only necessary for the x11 fallback. Though I'm guessing it might be best to keep it for now as the fallback can be useful.

I think this could bring another issue : Xorg support is, as of early 2020, still a necessity for some people that are stuck with non-updated software or with proprietary drivers that are not compatible with Wayland (like nvidia with `eglstreams`).

Reply:
I have some code laying around, something like this should do it IIRC:
```diff
modified   pkgs/desktops/gnome-3/core/gnome-session/default.nix
@@ -6,6 +6,8 @@ stdenv.mkDerivation rec {
   pname = "gnome-session";
   version = "3.34.2";
 
+  outputs = ["out" "sessions"];
+
   src = fetchurl {
     url = "mirror://gnome/sources/gnome-session/${stdenv.lib.versions.majorMinor version}/${pname}-${version}.tar.xz";
     sha256 = "1qgqp97f8k2zi2ydvhds14zsjzfj7cv521r6wx5sw0qacn0p7dwb";
@@ -39,6 +41,12 @@ stdenv.mkDerivation rec {
     patchShebangs meson_post_install.py
   '';
 
+  postInstall = ''
+    mkdir $sessions
+    moveToOutput share/wayland-sessions "$sessions"
+    moveToOutput share/xsessions "$sessions"
+  '';
+
   # `bin/gnome-session` will reset the environment when run in wayland, we
   # therefor wrap `libexec/gnome-session-binary` instead which is the actual
   # binary needing wrapping
```

> Xorg support is, as of early 2020

Right, with the X11 fallback I'm referring to the session gnome falls back to if it fails to launch through systemd (which is what happens with GDM without gnome3.enable). I see that wasn't very clear.

But yeah, I think just having another output is best, as the fallback might be useful.